### PR TITLE
CP-46806: Add master's cert thumbprint to header when HOST_IS_SLAVE

### PR DESCRIPTION
**Pull request by snwoods**

### CODE REVIEW

**Comment by snwoods** [comment]
Created as a draft as I still need to create unit tests but thought I'd open it up to feedback first.

File: ocaml/xapi/api_server.ml [hunk 1]
```diff
@@ -237,6 +237,30 @@ let callback1 ?(json_rpc_version = Jsonrpc.V1) is_json req fd call =
 
 (* debug(fmt "response = %s" response); *)
 
+let is_host_is_slave_error (response : Rpc.response) =
+  match response.contents with
+  | Rpc.(Enum [String x; String _]) when x = "HOST_IS_SLAVE" ->
```
Comment (robhoes) [inline ocaml/xapi/api_server.ml:242]: Use `Api_errors.host_is_slave`.

Comment (robhoes) [reply ocaml/xapi/api_server.ml:242]: Also, there must be a way to match on the failure condition in the `response`?

File: ocaml/xapi/api_server.ml [hunk 2]
```diff
@@ -237,6 +237,30 @@ let callback1 ?(json_rpc_version = Jsonrpc.V1) is_json req fd call =
...
+      true
+  | _ ->
+      false
+
+let create_thumbprint_header req response =
+  let cert_thumb_string = "x-xenapi-request-host-certificate-thumbprint" in
```
Comment (robhoes) [inline ocaml/xapi/api_server.ml:248]: Let's put the header name and value in xapi_globs.

**Review by robhoes — APPROVED** [review]
I think this looks good now (assuming it works :) )

File: ocaml/xapi/api_server.ml [hunk 3]
```diff
@@ -237,6 +237,33 @@ let callback1 ?(json_rpc_version = Jsonrpc.V1) is_json req fd call =
 
 (* debug(fmt "response = %s" response); *)
 
+let is_host_is_slave_error (response : Rpc.response) =
+  match response.contents with
+  (* RPC response returned by the rpc endpoint *)
+  | Rpc.(Enum [String x; String _]) when x = Api_errors.host_is_slave ->
+      true
+  (* RPC response returned by the jsonrpc endpoint *)
```
Comment (robhoes) [inline ocaml/xapi/api_server.ml:245]: When this function is called, we have't yet converted the response to XML or JSON, so why is this case needed?

**Review by mg12 — APPROVED** [review]

File: ocaml/xapi/api_server.ml [hunk 4]
```diff
@@ -259,13 +291,16 @@ let callback is_json req bio _ =
...
-          (Http.Hdr.content_type, "text/xml")
-        ; ("Access-Control-Allow-Origin", "*")
-        ; ("Access-Control-Allow-Headers", "X-Requested-With")
-        ]
+        ([
+           (Http.Hdr.content_type, "text/xml")
+         ; ("Access-Control-Allow-Origin", "*")
+         ; ("Access-Control-Allow-Headers", "X-Requested-With")
+         ]
+        @ thumbprint_header
```
Comment (psafont) [inline ocaml/xapi/api_server.ml:302]:
```suggestion
        ((Http.Hdr.content_type, "text/xml")
        :: ("Access-Control-Allow-Origin", "*")
        :: ("Access-Control-Allow-Headers", "X-Requested-With")
        :: thumbprint_header
```

File: ocaml/xapi/api_server.ml [hunk 5]
```diff
@@ -293,17 +328,20 @@ let jsoncallback req bio _ =
...
-          (Http.Hdr.content_type, "application/json")
-        ; ("Access-Control-Allow-Origin", "*")
-        ; ("Access-Control-Allow-Headers", "X-Requested-With")
-        ]
+        ([
+           (Http.Hdr.content_type, "application/json")
+         ; ("Access-Control-Allow-Origin", "*")
+         ; ("Access-Control-Allow-Headers", "X-Requested-With")
+         ]
+        @ thumbprint_header
```
Comment (psafont) [inline ocaml/xapi/api_server.ml:343]: ``suggestion
        ((Http.Hdr.content_type, "application/json")
        :: ("Access-Control-Allow-Origin", "*")
        :: ("Access-Control-Allow-Headers", "X-Requested-With")
        :: thumbprint_header
```

**Pull request merged**